fix(ui): delay win_viewport until the screen is updated

fredizzimo opened the pull request:
# Problem

Sometimes, when nvim sends the `win_viewport` event, for example when scrolling with visible folds on the screen, it reports the `scroll_delta` value one batch into "future". So when the client application is trying to show the new viewport it's not yet updated and results temporary corruption and flickering of the screen.

For a more detailed description see #23609, and starting from [this comment]( https://github.com/neovide/neovide/pull/1790#issuecomment-1518697747) in https://github.com/neovide/neovide/pull/1790,, where the issue was first detected. Note that some of the conclusions in those are not fully accurate, but the general observations are.

The documentation is also a bit unclear, especially regarding when the application is allowed to make updates to the new viewport.

# Solution
When there are pending updates to a Window, the `win_viewport` UI event is delayed until the updates are sent. This ensures that there's no flush between sending the viewport and updating of the lines corresponding to the new viewport.

The documentation has been clarified to document the existing behaviour (in the cases where there were no extra flushes) of the viewport, with an example of how applications can deal with the slightly surprising behaviour of the viewport event being sent after the updates.

Fixes: https://github.com/neovim/neovim/issues/23609

fredizzimo committed:
fix(ui): delay win_viewport until the screen is updated

When there are pending updates to a Window, the win_viewport UI event is
delayed until the updates are sent. This ensures that there's no flush
between sending the viewport and updating of the lines corresponding to
the viewport.

Previously that happened, for example when scrolling up and down and
some of the lines were folded. But it could happen any time
w_viewport_invalid is set and ui_flush was called.

fredizzimo commented:
I'm not sure how to test this.

`Screen:except` only takes a timeout, and I need to check that the state is as expected after the very next `flush`. Eventually it will become correct even without this changes, it's just that the intermediate changes are wrong.

clason commented:
> Another thing I'm not sure about is the CI failure. All passed on my local fork

CI is rather flakey these days, unfortunately. I'd recommend ignoring single(!) `test/` failures while the PR is still in draft. (Afterwards I can rerun tests to check for flake vs. consistent failure.)

fredizzimo committed:
docs(ui): clarify the win_viewport behavior

fredizzimo commented:
I added some clarification to the `win_viewport` documentation and removed the draft status. 

I tried to search for some ways of unit testing it, but it does not appear to be easy currently, so I left the tests out. But I will add some if you have any suggestions. But at least the is `disabled in folded text` test seems o have the correct behavior now. I don't think the column should be `9`, since the cursor is at the start of the search pattern.

zeertzjq commented on runtime/doc/ui.txt:
```diff
@@ -636,6 +636,13 @@ tabs.
 
+	All updates, such as `grid_line`, in a batch affects the new viewport,
+	despite the fact that `win_viewport` is received after the updates.
+	Applications implementing for example, smooth scrolling should take
```
```suggestion
	Applications implementing, for example, smooth scrolling should take
```

fredizzimo replied:
This should be fixed now

zeertzjq commented on src/nvim/drawscreen.c:
```diff
@@ -629,7 +629,12 @@ int update_screen(void)
+
+      int win_update_type = wp->w_redr_type;
       win_update(wp, &providers);
+      if (win_update_type) {
```
Is this check needed? This is only reached when `wp->w_redr_type != 0`

fredizzimo replied:
Good catch! No it appears like it isn’t. I must have missed the outer block that already does the check, I’m not used two reading code with just two space indentation.

fredizzimo replied:
This should be fixed now

justinmk commented on src/nvim/window.c:
```diff
@@ -1027,9 +1027,11 @@ void ui_ext_win_position(win_T *wp, bool validate)
 }
 
-void ui_ext_win_viewport(win_T *wp)
+void ui_ext_win_viewport(win_T *wp, bool allow_updated)
```
would it make senes to use the `validate` pattern found in many other display functions (including `win_ui_flush` which calls this)? Or are the semantics not similar enough?

```suggestion
void ui_ext_win_viewport(win_T *wp, bool validate)
```

justinmk commented on src/nvim/drawscreen.c:
```diff
@@ -629,7 +629,9 @@ int update_screen(void)
       }
+
       win_update(wp, &providers);
+      ui_ext_win_viewport(wp, true);
```
This placement seems kind of random. If `win_ui_flush` passed its `validate` arg to `ui_ext_win_viewport` , would this still be needed?

fredizzimo replied:
That would be too early. The viewport has to be sent after `win_update`, since that's the function that modifies the grid and indirectly calls `grid_line`

And there would be the same problem as before, that the grid is invalid at the new viewport, even after `flush` has been received.

justinmk reviewed (commented):
need @MDeiml @bfredl to review

# Problem

From https://github.com/neovim/neovim/issues/23609#issuecomment-1555187440 :

> When nvim sends `win_viewport` event, it reports the `scroll_delta` value a few batches into "future". I.e. the `scroll_delta` value is for a screen state that is only transferred after a few more batches (a few `flush` events) later.

From https://github.com/neovim/neovim/issues/23609 : 

> the updates are not atomic when scrolling, this is an example of two batches of UI events (for a slightly more detailed and different explanation see [neovide/neovide#1790](https://github.com/neovide/neovide/pull/1790)) : 
>
> [example data]
>
> the first batch sends a flush with a new viewport. But the problem here is that the viewport has not yet been updated, and Neovide has no way of knowing that, so it just displays what was on the screen at that location. The viewport is then redrawn in the second event. So, the whole sequence looks like the screen is flickering due to displaying the wrong contents temporarily.
>
> The real problem here is that the win_viewport event is underspecified. 

# Solution

When there are pending updates to a window, delay the `win_viewport` UI event until the updates are sent. This ensures there's no flush between sending the viewport and updating the viewport lines.

fredizzimo commented:
I still noticed problems in Neovide  when setting the new Neovim `smoothscroll` option. I have not debugged it completely yet, but it looks like the problem is that `scroll_delta` shows the wrong number of lines. It should always be 1 when I set `wrap` and `smoothscroll` and scroll with `<C-e>`. But it's sometimes 2 and sometimes 0.

zeertzjq commented on src/nvim/window.c:
```diff
@@ -1027,9 +1027,11 @@ void ui_ext_win_position(win_T *wp, bool validate)
 }
 
-void ui_ext_win_viewport(win_T *wp)
+void ui_ext_win_viewport(win_T *wp, bool allow_updated)
```
Is this argument really needed? If the window needs another redraw after calling `win_update()`, shouldn't `win_viewport` still be delayed?
```suggestion
void ui_ext_win_viewport(win_T *wp)
```

fredizzimo replied:
It is unfortunately required. `win_viewport` has the dual duty of indicating the row and column, and that is allowed to happen without calling `redraw_later` In those cases `v_viewport_invalid` is set and `ui_ext_win_viewport` is called directly from `UIFlush` and can happen basically everywhere.

So, if the check `allow_updated` is not there, we are back to square one, `win_viewport` gets sent with a flush, while the actual screen is still dirty.

My first approach was actually to remove `win_viewport` from `UIFlush`, but then there will be no feedback when only the `curcolumn` or `curow` are updated. The next approach was to call `win_view_port` unconditionally after `win_update`, but some tests are still failing with that, and it also sounds extremely risky, since it's been called from so many places before, even outside the `update_screen`

justinmk replied:
> `win_viewport` has the dual duty of indicating the row and column, and that is allowed to happen without calling `redraw_later` In those cases `v_viewport_invalid` is set and `ui_ext_win_viewport` is called directly from `UIFlush` and can happen basically everywhere.

This would be useful to have as a `@param` docstring.

zeertzjq commented on src/nvim/window.c:
```diff
@@ -1027,9 +1027,11 @@ void ui_ext_win_position(win_T *wp, bool validate)
-  if ((wp == curwin || ui_has(kUIMultigrid)) && wp->w_viewport_invalid) {
+  if ((wp == curwin || ui_has(kUIMultigrid))
+      // Delay the viewport command until the screen is updated if there are pending modifications
+      && (allow_updated || (wp->w_viewport_invalid && wp->w_redr_type == 0))) {
```
```suggestion
      && (wp->w_viewport_invalid && wp->w_redr_type == 0)) {
```

fredizzimo replied:
See my previous comment about `allow_updated`

zeertzjq commented on src/nvim/window.c:
```diff
@@ -1027,9 +1027,11 @@ void ui_ext_win_position(win_T *wp, bool validate)
 {
-  if ((wp == curwin || ui_has(kUIMultigrid)) && wp->w_viewport_invalid) {
+  if ((wp == curwin || ui_has(kUIMultigrid))
+      // Delay the viewport command until the screen is updated if there are pending modifications
```
```suggestion
      // Delay win_viewport until the screen is updated if there are pending modifications
```

fredizzimo replied:
I clarified the whole comment

bfredl commented on runtime/doc/ui.txt:
```diff
@@ -636,6 +636,13 @@ tabs.
 	purpose it only counts "virtual" or "displayed" lines, so folds
 	only count as one line.
 
+	All updates, such as `grid_line`, in a batch affects the new viewport,
+	despite the fact that `win_viewport` is received after the updates.
+	Applications implementing, for example, smooth scrolling should take
+	this into account and keep the grid separated from what's displayed on
+	the screen and copy it to the viewport destination once `win_viewport`
+	is received.
```
Is the `win_viewport` still sent as part of the _same_ batch (logical batch, i e before the final "flush" event)? Then perhaps this specific callout is redundant or should be toned down a little. State between different objects is not fully synchronized until "flush" is received, this is just how all screen events work.

If the concern is _mainly_ the space-bar heating change of changing the order to what some existing UIs might have come expect, then it is better to call out such a change in `news.txt` rather than cluttering the docs.

fredizzimo replied:
The documentation is the first place I go looking for this kind of information, and it also concerns people starting to make new UIs, or trying to fix bugs with their old information.

This would not be needed, if we somehow can change the order so that `win_viewport` comes before the `grid_scroll` and `grid_line` calls in the batch. But as it is now, coming the last in the batch, applications are forced to do some kind of double buffering.

Intuitively, users will think that things are sent in the correct order, and that flush is just a point where everything is consistent.

bfredl commented:
I'd structure the logic somewhat differently. `ui_ext_win_viewport()` should not take a `bool actually_do_it` parameter. Either the function is called in order to do its thing, or the function is not called.

instead, we should inhibit this function to be called when it is too early. Concretely, `win_ui_flush()` needs to take a `bool viewports` parameter, which should be set to `false` when called from `update_screen()` . alternatively,  this loop should be broken of the function entirely and still be done as part of `ui_flush()`

fredizzimo commented:
@bfredl,

As far as I can see, that needs a quite big restructuring of the code, not something I want to do. First of all there's no `win_ui_flush` in `update_screen` that is called after the grid is redrawn. The actual flush for that update only seems to happen the next frame.

And that one happens in the middle of the `update_screen` function, there has to be a reason for that, so I did not even attempt to move it to the end.

Secondly `win_ui_flush` is called from a huge amount of different places, some of them has to send `win_viewport` in order to update `curcol` and `currow`.

Yes, we could maybe add another `ui_flush` at the end of `update_screen`, and make that the only one to send `win_viewport`. But what consequences does an extra flush have here? Is `draw_screen` guaranteed to be called when just the cursor moves? 

But I just now did a quick test by moving `win_ui_flush` to the end of `draw_screen` and use only that as a signal to call `win_viewport` and the tests started to fail because `cur_col` is not updated properly.

bfredl commented:
> As far as I can see, that needs a quite big restructuring of the code, not something I want to do. 

I don't think so.

First of all there's no win_ui_flush in update_screen that is called after the grid is redrawn.

I didn't mean you should remove the new calls to `ui_ext_win_viewport(wp, true)`, just removing it unconditionally from the win_ui_flush that happens before.

fredizzimo commented:
@bfredl, thanks, 

It doesn't quite work like, that but brought me on the right track to simplify it. Starting with the original code, the only change needed is this to never send `win_viewport` when it's dirty. So no additional calls to `ui_ext_win_viewport` or anything

```diff
diff --git a/src/nvim/window.c b/src/nvim/window.c
index 8c2d2e3da..7f9126e8c 100644
--- a/src/nvim/window.c
+++ b/src/nvim/window.c
@@ -1029,7 +1029,7 @@ void ui_ext_win_position(win_T *wp, bool validate)
 
 void ui_ext_win_viewport(win_T *wp)
 {
-  if ((wp == curwin || ui_has(kUIMultigrid)) && wp->w_viewport_invalid) {
+  if ((wp == curwin || ui_has(kUIMultigrid)) && wp->w_viewport_invalid && wp->w_redr_type == 0) {
     int botline = wp->w_botline;
     int line_count = wp->w_buffer->b_ml.ml_line_count;
     if (botline == line_count + 1 && wp->w_empty_rows == 0) {
```
Of course this whole check could additionally be moved out to `win_ui_flush` since that's the only caller now. If you prefer it that way.

All tests passes with just this change, and it also appears to be working correctly with Neovide.

However, if I apply the test change that I made in this PR, it fails. I believe that change is the correct behaviour, but moving the `win_ui_flush(true)` inside `update_screen` to the end of the function fixes that and everything still works.

justinmk commented:
> but moving the `win_ui_flush(true)` inside `update_screen` to the end of the function fixes that and everything still works.

Nice. Updating the PR will help with review.

fredizzimo commented:
> Nice. Updating the PR will help with review.

Yes, I will update it later today. I was just waiting for confirmation that it's the correct approach to take.

fredizzimo commented:
I simplified the changes, and removed the changes to the `search highlighting` test. When I tested it earlier I thought that moving `win_ui_flush` to the end of `update_screen` would make it behave the same way as earlier, but I must have done something wrong because some other tests start to fail.

Adding an extra `win_ui_flush`, rather than just moving it to the end of `update_screen` does make it behave the same way. But I don't think that's the right way to fix the issue, we are probably missing a `w_viewport_invalid = false` somewhere and that's a separate issue. 

I have not done any new changes to the documentation yet, since it's unclear to me how it should be changed. `win_viewport` still works the same as before, it comes last in the batch, which is at least to me surprising, since all other events can be processed in the order they are received, as long as the only displayed state is the one right after flush.

Note, that I have not yet tested these latest changes very much, only done a few sanity checks and that it indeed fixes the problem in Neovide. The `smoothscroll` bug I reported earlier here, is also fixed, probably with https://github.com/neovim/neovim/pull/24234

justinmk commented:
Is the Problem/Solution framing in https://github.com/neovim/neovim/pull/24182#pullrequestreview-1509525545 accurate or can you update the PR description with a correct Problem/Solution framing?

fredizzimo commented:
@justinmk, it was mostly correct, but had a few inaccuracies, like "a few batches" instead of "one batch",  so I rewrote it and updated the description.

fredizzimo commented:
@justinmk, I was referring to comment @bfredl gave earlier here https://github.com/neovim/neovim/pull/24182#discussion_r1250390961. From my point of view it's good as it is, but I can change it, if you prefer to not have that many details there. For example by removing the last sentence:
> Applications implementing, for example, smooth scrolling should take this into account and keep the grid separated from what's displayed onthe screen and copy it to the viewport destination once `win_viewport` is received.

fredizzimo commented:
@zeertzjq, @justinmk. 

Would it be possible to backport this into the next 0.9.x release? We just ran into some problems here https://github.com/neovide/neovide/pull/1870#issuecomment-1647863757, which are hard to fix  without breaking things completely when using it  with 0.9.1 and older. Therefore, I think we have to limit the version of Neovim people can use together with it.

0.10, is probably quite far away, so including it in 0.9.2 would make things easier for us. I don't think this is a breaking change for existing clients, but I totally understand if you don't want to include it.

vhakulinen commented:
Gnvim is using this at the moment (for smooth scrolling) and if I've understood this change correctly it will break, albeit already broken, behavior. I can test it later this week.

fredizzimo commented:
I just double checked Neovide release, which does not take advantage of the fixes. And there it works the same with Neovim nightly and 0.9.1. But that's just one data point. I will check if someone on our side can test all major GUI implementations to see if there are any breakages.

justinmk commented:
> Would it be possible to backport this into the next 0.9.x release?

Can you open a PR against the `release-0.9` branch?

fredizzimo commented:
@justinmk, yes here https://github.com/neovim/neovim/pull/24480

vhakulinen commented:
> Gnvim is using this at the moment (for smooth scrolling) and if I've understood this change correctly it will break, albeit already broken, behavior. I can test it later this week.

~Later this week~ one month later I managed to get back to this. Required changes for gnvim were minimal and the scrolling works better than ever. Neovim sends some wrong delta values from time to time tho'.
